pkg/backup/item_block_worker_pool.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright the Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"context"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type ItemBlockWorkerPool struct {
+	inputChannel chan ItemBlockInput
+	wg           *sync.WaitGroup
+	logger       logrus.FieldLogger
+	cancelFunc   context.CancelFunc
+}
+
+type ItemBlockInput struct {
+	itemBlock  *BackupItemBlock
+	returnChan chan ItemBlockReturn
+}
+
+type ItemBlockReturn struct {
+	itemBlock *BackupItemBlock
+	resources []schema.GroupResource
+	err       error
+}
+
+func (p *ItemBlockWorkerPool) GetInputChannel() chan ItemBlockInput {
+	return p.inputChannel
+}
+
+func StartItemBlockWorkerPool(ctx context.Context, workers int, log logrus.FieldLogger) *ItemBlockWorkerPool {
+	// Buffer will hold up to 10 ItemBlocks waiting for processing
+	inputChannel := make(chan ItemBlockInput, 10)
```

Comment:
Why do we pick 10 as the hardcode value, instead of number of CPU cores?

Reply:
CPU cores isn't relevant here. This is just the buffer -- the number of not-yet-taken input values that can be added to the channel before the adding process blocks. The default value for the number of goroutines processing inputs will be one.

Reply:
I am just not very confident with the number of 10. Since we have <pool size> workers, why do we only allow 10 inputs suppose that <pool size>  > 10?
In future, when the logics are changed, not sure we can easily recall this number which may be any bottleneck because of logic change.

Probably, if we don't need a different configuration which may block the threads necessarily, we don't add it.

Reply:
so that I understand, are you advocating for no buffer in the channel or just that we don't know the correct value right now and it should be configurable or based on the number of workers?

Reply:
Pool size and buffer size are completely unrelated to each other. Removing the buffer completely will increase blocking on the input of the channel, so I think we definitely want a buffer. The only question here is whether a hard-coded value of 10 is sufficient for all cases or whether we want to make this user-configurable. I don't have a strong objection to adding another install/server flag for this, I just don't know if 1) there's any need for it or 2) whether it will confuse users.

Reply:
I think the goal here is to remove an unnecessary point in the logic that could block the workflow.
So I think there are two options:
1. Set it with a very large number, the same as the progress update channel
2. Set it the same as the work pool size or several folds of the pool size. Though the pool size and the buffer size are unrelated, setting it based on the pool size ensures that we no worker is starve at any point of time

In this way, users never need to tune this separately, so it doesn't need to be configurable.

Reply:
This should be fine, thanks.

Reply:
This should be fine, thanks.